app/models/role.rb

```diff
@@ -129,7 +137,39 @@ def add_permissions!(*args)
     save!
   end
 
-private
+  def set_taxonomies(user_selected_ids)
+    user_selected_ids = user_selected_ids.reject(&:blank?).map(&:to_i)
+    existing_taxonomies = existing_taxonomy_ids
+    user_can_set = User.current.my_taxonomies_ids
+
+    if (existing_taxonomies & user_can_set).sort != user_selected_ids.sort
+      # we want to keep associations that user can't modify
+      new_objects = (existing_taxonomies - user_can_set) + (user_selected_ids & user_can_set)
+      self.organization_ids = Organization.where(:id => new_objects).pluck(:id)
```

Comment:
I don't want to add extra metaprogramming only because these two objects are similar so I rather added `if Taxonomy.organizations_enabled?` which should have the same impact

btw you wanted me to save 0.00001s on `sort!` so I suppose the performance is also good argument here, enabled_taxonomies creates new array instance, runs select + all the code that would construct "organization_ids=" input for send...